src/main/scala/piconot/PicoFSM.scala

```diff
@@ -0,0 +1,98 @@
+package piconot
+
+import scalafx.application.JFXApp
+import picolib.maze.Maze
+import picolib.semantics
+import scala.collection.mutable.MutableList
+import scala.language.experimental.macros
+
+trait PicoBot extends JFXApp {
+
+  var maze: Maze = null;
+  val rules: MutableList[semantics.Rule] = new MutableList
+  var curState : State = null
+
+  object use {
+    def map(path: String) = {maze = Maze(path)}
+  }
+  //def map(path: String) = {maze = Maze(path)}
+
+  class Direction(val dir : semantics.MoveDirection) {
+    def unary_+ = { new EnvList(List(this), List()) }
+    def unary_- = { new EnvList(List(), List(this)) }
+  }
+
+  class CanMakeRule(posDirs : List[Direction],
+                    negDirs : List[Direction]) {
+
+    object ConstraintConflictException extends Error
+
+    def makeRule(moveDir : semantics.MoveDirection, nextState : State) = {
+      def getBlocked(d: Direction) = {
+        val inN = negDirs contains d
+        val inP = posDirs contains d
+
+        if (inN && inP) throw ConstraintConflictException
```

Comment:
I like that you do error checking! This is something we completely ignored/forgot.